### PR TITLE
update width calculations

### LINKED ISSUE #6: Fill up rows as much as possible in `LeftToRight` direction
Below is a test adapted from the test in https://github.com/ogham/rust-term-grid/pull/15. It currently fails in this fork too.

```rust
#[test]
fn theoretical_is_effective_max_num_line() {
    let mut grid = Grid::new(GridOptions {
        filling: Filling::Text("||".into()),
        direction: Direction::LeftToRight,
    });

    for s in [
        "test1", "test2", "test3", "test4", "test5", "test6", "test7", "test8", "test9", "test10",
        "test11",
    ] {
        grid.add(Cell::from(s));
    }

    let bits =
        "test1 ||test2 ||test3||test4||test5||test6||test7||test8||test9\ntest10||test11||\n";

    let display = grid.fit_into_width(69).unwrap();
    assert_eq!(display.to_string(), bits);
    assert_eq!(display.row_count(), 2);
}
```

Expected behaviour:
```
test1 ||test2 ||test3||test4||test5||test6||test7||test8||test9
test10||test11||
```

Actual behaviour:
```
test1||test2||test3||test4 ||test5 ||test6
test7||test8||test9||test10||test11||
```

This expected behaviour matches the behaviour of GNU `ls`:
```
❯ ls --sort=version -x
test1   test2   test3  test4  test5  test6  test7  test8  test9
test10  test11
```

However, this behaviour changes in the column format:
```
❯ ls --sort=version
test1  test3  test5  test7  test9   test11
test2  test4  test6  test8  test10
```

Note how it's not:
```
test1  test3  test5  test6 test7  test8  test9  test10  test11
test2  test4
```

We have three options to proceed:
- Settle on one behaviour and document the difference with GNU in `uutils/coreutils`.
- Make this behaviour configurable regardless of direction.
- Make `Direction::LeftToRight` fill up as much of the lines and make `Direction::TopToBottom` fill up columns.

I'd like to coordinate also with the `exa` folks to determine whether they want to use this fork too and what behaviour they want.

### PR DESCRIPTION
**Pull request by emar-kar**
Closes #6 

I was trying to improve performance in grid calculation and switch from rows to columns to fix the related issue

### CODE REVIEW

**Comment by emar-kar** [comment]
@tertsdiepraam here is the MR I was talking about. And to update `ls` do I'll wait for the release of the library.

**Comment by tertsdiepraam** [comment]
Uh oh, looks like I was just touching the same code 😄 I'll look at this tomorrow.

**Comment by tertsdiepraam** [comment]
I'm frankly amazed this works for `TopToBottom` too. I was just it didn't but it works because we don't reference `num_columns` in the `compute_dimensions` for `TopToBottom`. It's a bit strange. I'm gonna have to poke at this for a bit, because I think there will be some strange edge cases. This is always such hard code to reason about fully.

**Comment by emar-kar** [comment]
It calculates the number of possible columns first using the terminal width and widest possible cell content. I thought it's more relevant, since both of the values are given and it can be the "worst" case scenario e.g. all names are the size of widest. Then it tries to increase the number of columns and see if the `potential_dimension`, which calculates columns widths according to print direction still fits into the terminal width.

> I was just it didn't but it works because we don't reference `num_columns` in the `compute_dimensions` for `TopToBottom`.

`compute_dimensions` calculates the width of the cells according to the direction and it does not use `num_columns` in `TopToBottom` because it is the index of the element in cells vector.

If you can come up with some edge cases I didn't think about it would be great, thx

File: src/lib.rs [hunk 1]
```diff
@@ -179,94 +177,52 @@ impl<T: AsRef<str>> Grid<T> {
...
+        // Try to increase number of columns, to see if new dimension can still fit.
+        for num_columns in appr_num_columns + 1.. {
+            let new_width = self.options.width - (num_columns - 1) * self.options.filling.width();
+            let num_lines = div_ceil(self.cells.len(), num_columns);
+            let new_dimension = self.compute_dimensions(num_lines, num_columns);
+            if new_dimension.widths.iter().sum::<usize>() <= new_width {
+                potential_dimension = new_dimension;
             } else {
                 break;
             }
```
Comment (tertsdiepraam) [inline src/lib.rs:222]: This break is not 100% correct. This test will fail with it. I've got a solution to that on my commit, but it will increase the computation time a bit again.
```rust
#[test]
fn weird_column_edge_case() {
    // Here, 5 columns fit while fewer columns don't. So if we exit too early
    // while increasing columns, we don't find the optimal solution.
    let grid = Grid::new(
        vec!["0", "1", "222222222", "333333333", "4", "5", "6", "7", "8"],
        GridOptions {
            direction: Direction::TopToBottom,
            filling: Filling::Spaces(2),
            width: 21,
        },
    );

    let expected = "\
        0  222222222  4  6  8\n\
        1  333333333  5  7\n\
    ";

    println!("{grid}");
    assert_eq!(grid.to_string(), expected);
}
```

File: src/lib.rs [hunk 2]
```diff
@@ -179,94 +177,52 @@ impl<T: AsRef<str>> Grid<T> {
...
-            // also serves as a speed-up.
-            let total_separator_width = (num_columns - 1) * self.options.filling.width();
-            if maximum_width < total_separator_width {
-                continue;
-            }
 
-            // Remove the separator width from the available space.
-            let adjusted_width = maximum_width - total_separator_width;
+        // Calculate number of columns with widest column size.
+        let max_num_columns = self.options.width / widest_column;
```
Comment (tertsdiepraam) [inline src/lib.rs:200]: This confused me for a long time, because I took the variable name at face value, but it's actually more the _minimum_ number of columns, or rather, if you `min` this with the number of cells it's the minimum. Like this:
```rust
        let min_columns = self
            .cells
            .len()
            .min((self.options.width + self.options.filling.width()) / widest_column);
```
I think that removes the need for the `appr_num_columns` calculation below.

**Review by tertsdiepraam — CHANGES_REQUESTED** [review]
Right! I've looked _much_ more into this now and I think I have a better grasp of what you've done. In short: it's fantastic! However, it does need a lot of improvements in terms of variable names and comments. I've got a commit where I was doing that for myself, so with your permission I'd like to push that to this branch. I also included more tests.

There are some other issues. The `is_complete` and `width` functions are now broken because columns can now have 0 widths in some columns for `TopToBottom`. We should investigate whether uutils and `eza` use those (I don't think so), so we can remove them.

I think you've also solved https://github.com/uutils/uutils-term-grid/issues/39, because you're assuming balanced columns. Which seems to match what GNU is doing:
```
> ls --width=6
aaa
bbb
ccc
d
```
In current uutils-term-grid, this would be printed as
```
aaa  d
bbb
ccc
```
But that is unbalanced and kind of ugly. The reason I suppose that this was done in the first place is that it is technically "more optimal" in terms of line count.

All in all, consider me impressed! Do I have your permission to push to this branch?

**Comment by emar-kar** [comment]
@tertsdiepraam Yeah, sure. If you need to push something, you are more than welcome to

File: src/lib.rs (hunk 1 above)
Comment (emar-kar) [reply src/lib.rs:222]: Yeah, I see the point. That's exactly the edge case I didn't think about. But I think I come up with the solution to this problem. Can I push this test with the fix commit, so you can check it out?

Comment (tertsdiepraam) [reply src/lib.rs]: Sure! I'll rebase my changes on that then

Comment (emar-kar) [reply src/lib.rs]: Even tho it fixed the issue I just realized that I drastically increased the number of computations. I'll have a look into this part more later. But if you have any suggestions, just push those

**Comment by tertsdiepraam** [comment]
Whoo it finally works. For some reason git had a really hard time pushing to this branch. It's mostly updating some names and comments.

> Even tho it fixed the issue I just realized that I drastically increased the number of computations. I'll have a look into this part more later. But if you have any suggestions, just push those

I set a maximum number of columns of `self.cells.len()`, but otherwise I don't have any great ideas. How are you measuring the performance?

**Comment by tertsdiepraam** [comment]
There's actually another algorithm I was thinking of for the `LeftToRight` direction, which might speed things up again.

Basically we can just start building the first row until we hit the maximum width. This gives us a pretty good maximum number of columns. Then we just decrease the number of columns and calculate the dimensions each time until we find the first one that fits.

For the `TopToBottom` direction we can then maybe also be a bit smarter by specializing for it. We would need to split the computation for both. Maybe an idea for another PR? Then we'll merge this, then make a PR splitting the two directions and then we can optimize them separately.

**Comment by emar-kar** [comment]
> How are you measuring the performance?

I'm using `hyperfine` on **uu_ls** to see if the change of the crate does not affect performance there. But my take here was to reduce call to `compute_dimensions` as much as possible. 

> Basically we can just start building the first row until we hit the maximum width. This gives us a pretty good maximum number of columns.

I think this one can be tricky with edge cases. We can definitely try this approach to see if it will be more efficient than the one with using the max width.

> For the `TopToBottom` direction we can then maybe also be a bit smarter by specializing for it. We would need to split the computation for both. Maybe an idea for another PR? Then we'll merge this, then make a PR splitting the two directions and then we can optimize them separately.

Actually I was also thinking in this direction, to split the computations for every direction. It will grand more flexibility on improving individual parts, instead of coming up with something common for everything. Plus, there was a request to add `RightToLeft` direction. If we split these two now, IMHO it will be easier to add support for another the same way.

**Comment by tertsdiepraam** [comment]
I already have an idea to improve `TopToBottom` I think. Instead of checking each number of columns we do this:
```
current_rows = div_ceil(cells, current_cols)
next_rows = current_rows - 1
next_columns = div_ceil(cells, next_rows)
```
I _think_ (but have to check) that this will skip possibilities that lead to empty columns at the end

**Review by tertsdiepraam — APPROVED** [review]
Merging so we can make a new branch for splitting it up!
